File: android/app/src/main/java/com/synapse/messenger/network/protocol/Frame.kt
package com.synapse.messenger.network.protocol

import java.io.ByteArrayOutputStream
import java.util.zip.GZIPInputStream

/**
 * Frame codec — mirrors `server/pkg/wire/frame.go`.
 *
 * ```
 *  +--------+--------+--------+--------+--------------------+==================+
 *  | 0x53   | 0x43   | VER(1) | FLAGS  | LENGTH (4, BE)     | PAYLOAD (LENGTH) |
 *  |  'S'   |  'C'   |  0x01  | bits   | uint32 <= 16 MiB   | envelope bytes   |
 *  +--------+--------+--------+--------+--------------------+==================+
 * ```
 *
 * On the WebSocket transport one binary message carries exactly one frame, so
 * there is no stream re-assembly here — [decode] takes a whole message.
 *
 * Outbound frames are never compressed. The gateway does not require symmetry
 * (flags are per frame), and a phone's uplink carries short frames where gzip
 * would cost CPU for nothing. Inbound frames may be gzipped because we advertise
 * [Cap.COMPRESSION]; zstd cannot appear because we do not advertise it.
 */
object Frame {
    const val MAGIC_0 = 0x53 // 'S'
    const val MAGIC_1 = 0x43 // 'C'
    const val VERSION = 0x01
    const val HEADER_SIZE = 8

    /** Matches the server's parser bound; a larger length prefix is hostile input. */
    const val MAX_PAYLOAD_SIZE = 16 shl 20

    const val FLAG_NONE = 0
    const val FLAG_COMPRESSED = 1 shl 0 // gzip
    const val FLAG_ZSTD = 1 shl 1 // zstd + shared dictionary

    fun encode(payload: ByteArray, flags: Int = FLAG_NONE): ByteArray {
        require(payload.size <= MAX_PAYLOAD_SIZE) { "payload too large: ${payload.size}" }
        val frame = ByteArray(HEADER_SIZE + payload.size)
        frame[0] = MAGIC_0.toByte()
        frame[1] = MAGIC_1.toByte()
        frame[2] = VERSION.toByte()
        frame[3] = flags.toByte()
        val n = payload.size
        frame[4] = (n ushr 24).toByte()
        frame[5] = (n ushr 16).toByte()
        frame[6] = (n ushr 8).toByte()
        frame[7] = n.toByte()
        payload.copyInto(frame, HEADER_SIZE)
        return frame
    }

    /** Unwraps one complete frame and returns its (decompressed) envelope payload. */
    fun decode(frame: ByteArray): ByteArray {
        if (frame.size < HEADER_SIZE) throw FrameException("short frame: ${frame.size} bytes")
        if (frame[0].toInt() and 0xFF != MAGIC_0 || frame[1].toInt() and 0xFF != MAGIC_1) {
            throw FrameException("bad magic")
        }
        val version = frame[2].toInt() and 0xFF
        if (version != VERSION) throw FrameException("unsupported framing version $version, want $VERSION")

        val flags = frame[3].toInt() and 0xFF
        val length = ((frame[4].toInt() and 0xFF) shl 24) or
            ((frame[5].toInt() and 0xFF) shl 16) or
            ((frame[6].toInt() and 0xFF) shl 8) or
            (frame[7].toInt() and 0xFF)
        if (length < 0 || length > MAX_PAYLOAD_SIZE) throw FrameException("payload too large: $length")
        if (frame.size < HEADER_SIZE + length) throw FrameException("truncated frame")

        val payload = frame.copyOfRange(HEADER_SIZE, HEADER_SIZE + length)
        return when {
            flags and FLAG_ZSTD != 0 -> throw FrameException(
                "zstd frame received but CapZstd was never advertised",
            )
            flags and FLAG_COMPRESSED != 0 -> gunzip(payload)
            else -> payload
        }
    }

    private fun gunzip(compressed: ByteArray): ByteArray {
        val out = ByteArrayOutputStream(compressed.size * 3)
        GZIPInputStream(compressed.inputStream()).use { gz ->
            val buf = ByteArray(8 * 1024)
            var total = 0
            while (true) {
                val read = gz.read(buf)
                if (read <= 0) break
                total += read
                // Bound the output the same way the server does: a gzip bomb must not
                // become an OOM on a phone.
                if (total > MAX_PAYLOAD_SIZE) throw FrameException("decompressed payload too large")
                out.write(buf, 0, read)
            }
        }
        return out.toByteArray()
    }
}

class FrameException(message: String) : Exception(message)

File: client/src/shared/api/protocol/frame.ts
/**
 * Frame codec — mirrors `server/pkg/wire/frame.go`.
 *
 *   +--------+--------+--------+--------+--------------------+==================+
 *   | 0x53   | 0x43   | VER(1) | FLAGS  | LENGTH (4, BE)     | PAYLOAD (LENGTH) |
 *   |  'S'   |  'C'   |  0x01  | bits   | uint32 <= 16 MiB   | envelope bytes   |
 *   +--------+--------+--------+--------+--------------------+==================+
 *
 * On the WebSocket transport one binary WS message carries exactly one frame,
 * so there is no stream re-assembly to do here — `decodeFrame` takes a whole
 * message. The magic word lets the server cheaply reject garbage; we validate
 * it on the way in for the same reason.
 */

export const MAGIC_0 = 0x53 // 'S'
export const MAGIC_1 = 0x43 // 'C'
export const VERSION = 0x01
export const HEADER_SIZE = 8

/** Matches the server's parser bound; a larger length prefix is hostile input. */
export const MAX_PAYLOAD_SIZE = 16 << 20

/** Frame flag bits. */
export const Flag = {
  NONE: 0,
  /** payload is gzip-compressed */
  COMPRESSED: 1 << 0,
  /** payload is zstd-compressed (with a shared dictionary) */
  ZSTD: 1 << 1,
} as const

export class FrameError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'FrameError'
  }
}

/**
 * Wraps an envelope payload in a frame.
 *
 * The web client never sets a compression flag: it advertises neither
 * CapCompression nor CapZstd during the handshake, so neither side compresses
 * (see `caps.ts` for why). Outbound frames are therefore always raw.
 */
export function encodeFrame(payload: Uint8Array, flags: number = Flag.NONE): Uint8Array {
  if (payload.length > MAX_PAYLOAD_SIZE) {
    throw new FrameError(`payload too large: ${payload.length} > ${MAX_PAYLOAD_SIZE}`)
  }
  const frame = new Uint8Array(HEADER_SIZE + payload.length)
  frame[0] = MAGIC_0
  frame[1] = MAGIC_1
  frame[2] = VERSION
  frame[3] = flags
  new DataView(frame.buffer).setUint32(4, payload.length, false) // big-endian
  frame.set(payload, HEADER_SIZE)
  return frame
}

/** Unwraps one complete frame and returns its envelope payload. */
export function decodeFrame(frame: Uint8Array): Uint8Array {
  if (frame.length < HEADER_SIZE) {
    throw new FrameError('short frame')
  }
  if (frame[0] !== MAGIC_0 || frame[1] !== MAGIC_1) {
    throw new FrameError('bad magic')
  }
  if (frame[2] !== VERSION) {
    throw new FrameError(`unsupported framing version: got ${frame[2]}, want ${VERSION}`)
  }
  const flags = frame[3]
  const view = new DataView(frame.buffer, frame.byteOffset, frame.byteLength)
  const length = view.getUint32(4, false)
  if (length > MAX_PAYLOAD_SIZE) {
    throw new FrameError('payload too large')
  }
  if (frame.length < HEADER_SIZE + length) {
    throw new FrameError('truncated frame')
  }
  if (flags & (Flag.COMPRESSED | Flag.ZSTD)) {
    // Unreachable while we advertise no compression capability: the gateway
    // only sets these flags for peers that negotiated them. Fail loudly rather
    // than hand a compressed blob to the envelope parser.
    throw new FrameError(
      'received a compressed frame but no compression capability was negotiated',
    )
  }
  return frame.subarray(HEADER_SIZE, HEADER_SIZE + length)
}
